File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

def check(s):
    stk = []
    for c in s:
        if c == "(":
            stk.append("(")
        elif c == "[":
            stk.append("[")
        elif c == ")":
            if stk and stk[-1] == "(":
                stk.pop()
            else:
                return "no"
        elif c == "]":
            if stk and stk[-1] == "[":
                stk.pop()
            else:
                return "no"
    
    if stk:
        return "no"
    else:
        return "yes"    
    
while True:
    line = sys.stdin.readline()
    if line[0] == ".": break
    
    print(check(line))

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <string>
#include <iostream>
#include <stack>

using namespace std;

string check(string line){
    stack<char> stk;
    for(char c : line){
        if(c == '(' || c == '[') stk.push(c);
        else if(c == ')'){
            if(!stk.empty() && stk.top() == '(') stk.pop();
            else return "no";
        }
        else if(c == ']'){
            if(!stk.empty() && stk.top() == '[') stk.pop();
            else return "no";
        }
    }
    
    if(stk.empty()) return "yes";
    else return "no";
}

int main(){
    while(true){
        string line;
        getline(cin, line);
        
        if(line[0] == '.') break;
        
        cout << check(line) << endl;
    }
    
    return 0;
}